### ISSUE TITLE
Updating from server Craft 3 fails; workaround, but...

### ISSUE THREAD

**Issue by narration-sd** [craftcms/cms#1992]
### Description
Using source composer update, push to server always reliable; doing Craft update from server always fails. I found how to get it to work, but requires intervention every time.

### Steps to reproduce

1. Have a version on source w/vendor of Craft 3, beta.NN or dev-develop as 3.0.0-beta.99
2. Push all including /vendor to server; run Updates to see which, as there are some
3. Click to updates -- time passes for downloads
4. Update fails with a call for help button: .git missing, usually for craft-cms itself
5. Try a full cleanup of source including Composer clear-cache; re-install, push
6.  Update fails immediately with #1990; fix that w/maintenance bit clear in db
7. Update again; time passes, receive same fail
8. Now, the workaround. ssh into server, composer update there.
9. Get messages now more clearly about  missing .gits, with option to repair them
10. When done, clear maintenance bit in db again
11. Run Update from Craft. Succeed.
12 Receive word of another beta.nn update, or just change composer.json on source back to dev-develop etc., composer update, push.
13. Get unusual fail complaining of _back_ update???
14. Fail again at server Update once again. Correct item number here, 13

You can see this isn't fun, and certainly for release coming. I may have missed that you are already going to fix it; some ghost of possible memory there?

### Additional info

PHP version	7.1.3-3+deb.sury.org~xenial+1
Database driver & version	MySQL 5.7.17-0ubuntu0.16.04.1
Image driver & version	Imagick 3.4.3RC2, ImageMagick 6.8.9-9
Craft edition & version	Craft Pro, was 3.0.0-beta.26 this time
Yii version	2.0.12
Twig version	2.3.2
Guzzle version	6.2.1
Imagine version	0.7-dev
Plugins

Bactrian	1.0.0
Trimmer	1.0.0
Video Embed	3.0.0
Element API	2.4.3
ImageOptimize	1.2.2
Standards Agent	2.0.1

**Comment by narration-sd** [comment]
Here are the images of the fixup process from server ssh point of view, and of the second 'unusual' fail when futher updating aftwerwards, claiming we are backtracking, rather than the true case of moving forward beta.26->beta.28 via composer.json altered on server -- no source vendor push upload.

![fixup-log](https://user-images.githubusercontent.com/247945/30517641-c9cfef0a-9b1a-11e7-95c0-c55e1e36cb01.png)

![update-trace](https://user-images.githubusercontent.com/247945/30517649-f700ee98-9b1a-11e7-8c38-4c2a2474c009.png)

**Comment by brandonkelly** [comment]
Are you setting any custom `repositories` settings in your `composer.json` file?

**Comment by narration-sd** [comment]
yes, a bunch of file path repos for plugins  plus one  remote for Doxter.  I seem to remember the path ones causing a problem last time I tried the CP update, but the visible message was specific to them then, which it's not this time. 

In fact, that's why I tried this again in the first place; was intending to work that out.

**Comment by narration-sd** [comment]
...looking again at the second fail, and carefully not disturbing what I said I did there, might understand the message. It could make sense if what I actually did was only to update the composer.json and push that; with no local composer update command issued.

Then the  server would have earlier code (i.e. beta.26) code present in vendor, but a composer.json asserting it was dev-develop as beta.99. Thus seeming to have attempted a backing down. 

If that works for you...

**Comment by narration-sd** [comment]
Ok, I may understand what's going on. If so, it's because I initially had a dev-develop version on the server.

This problem we're seeing is known, and in fact the 'would you like to retry installing the package instead'is the sort-of fix they put in.  Refs here: 
- https://github.com/composer/composer/issues/4717
- https://github.com/composer/composer/pull/4878

So looks like it revolves around hiding the .git folder dev-develops have from Packagist.

I tried a manual update after Craft update failed similarly on another site, Attached images show a log of that, and a find where the .git foders appear after success.

The first ref may suggest why your dev-develops get the error. Why some symphony parts also? Maybe the ones that come with dev-develop have the same issue?

dev-develops, and/or pull requests merged locally on them like mine that you still have outstanding seem likely to occur in future. I can just continue updating on development machine and pushing all for updates, but when it comes time for production, what to do? 

I guess I could revert the dev machine to a non-dev-develop Craft then, but have you a better idea to work the problem?


![manual-update-yes](https://user-images.githubusercontent.com/247945/30531046-7d5cf474-9c00-11e7-93f3-8ae55f74cecc.png)

![dot-gits-present](https://user-images.githubusercontent.com/247945/30531049-85c33cf4-9c00-11e7-9df7-102d25c40cf6.png)

**Comment by brandonkelly** [comment]
Sounds like it has something to do with updating from _source_ (`dev-develop`) to _dist_ (`3.0.0-beta.X`).

Do you have [`config.preferred-install`](https://getcomposer.org/doc/06-config.md#preferred-install) set in your `composer.json` file by chance? If so please remove it.

Then delete your entire `vendor/` folder, and run

```
composer require craftcms/cms:"^3.0.0-beta.27"
```

**Comment by narration-sd** [comment]
Ok!  That's a good move.

> Do you have config.preferred-install set in your composer.json file by chance? If so please remove it.

No, didn't -- you had the composer.json to see above. Interesting directive though.

> `run composer require craftcms/cms:"^3.0.0-beta.27"`

Yes, that did the trick. Accepted, ran, Craft ended up properly on beta.28. I don't think I had to delete /vendor for this, and can cause issues with plugins depending on sources, but let's not get into that; understand why you would suggest it on this issue.

So, I went a little further. I did (notice strict version):

`run composer require craftcms/cms:"3.0.0-beta.27"`

Craft reporting then as asked on beta.27, and with one update. I ran that from the CP, which succeeded: now reporting again on beta.28, with no problems along the way.

I'm kind of turning over possible alternatives in mind, for how Craft might safety this overall issue -- or whether you would want to. 

If running a CP Craft update did the composer require to version itself, seems this would always work, and if you asked Craft to do it, that's what you'd want to happen, even if dev continues to require a dev-develop. 

That case would also its proper revert as soon as you uploaded again, for continuing dev, as this would presumably include dev's matching /vendor. Sound right?

I'm considering also the database, but again it seems this would come out right, because dev-develop on the dev machine would always be matched or ahead to the latest release on the server, no?

**Comment by narration-sd** [comment]
Hmm - and looks like I was wrong; you do have do delete /vendor, in addition to the composer require,  to succeed in getting around the 'missing .git' problem, the one that I started this issue about. Both steps are necessary.

Discovered working with a different installation, where I also found a fix for the plugins problem, which occurs whether you blow away /vendor or not. 

This was need to authenticate if you are holding some plugins in a private git/bitbucket facility. The solution there is again manual, but needed only once:  after you authenticate with login/passwd, composer soon offers to store those locally on the server, and this is used when composing again.

So I guess a certain amount of manual  jiggery-pokery is going to be the thing, around dev. And if you want to return a server to normal after dev-develop, I guess  you can do the steps as easily on the dev source machine, and upload the normalized /vendor at the same time as anything else. After which normal CP updates will work.

Last musing about this is concerning plugins, if they are not released by the Plugin Store, or before you want to do that, or given what features it may have for this. Say you are developing a private plugin for a client. They have access rights on a git provider. How do we get those into their Craft install? If I control that, I can with some fuss get the composer credentials just mentioned onto their server. If I don't control it, how? This part probably merits a separate issue, imagine you will think....

**Comment by narration-sd** [comment]
..and the quick conversion process, as well as assuring thte plugin repository access presence, can be part of the dev->master->tagged activity at release points. 

If you don't want better, I can probably write this up for doc.

**Comment by brandonkelly** [comment]
Glad you got it sorted. The CP updater is only focussed on simple use cases (moving between real releases, not development versions like `dev-develop`), so I’m not too worried about this becoming a common issue. But if it does end up being common, we can look into how we can make it more robust.

**Comment by narration-sd** [comment]
Sounds good, Brandon, thanks. Should you get to that, I think the info needed is in the ref links above; the .gits show up in the composer autoload
